### ISSUE TITLE
Refactor `transitive_closure` to support multigraphs

### ISSUE THREAD

**Issue by vdshk** [networkx/networkx#5043]
By definition (see below and #5016) `transitive_closure` also works with multigraphs.
* **transitive closure of G** is the graph G* = (V, E*), where E* = {(i, j) : there is a path from vertex i to vertex j in G}"

But now this function does not work as expected.

### Current Behavior
```python
>>> import networkx as nx
>>> g = nx.MultiDiGraph([(0, 1), (1, 2)])
>>> print(nx.transitive_closure(g).edges())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<class 'networkx.utils.decorators.argmap'> compilation 8", line 4, in argmap_transitive_closure_5
  File "/home/vdshk/.local/lib/python3.8/site-packages/networkx/algorithms/dag.py", line 576, in transitive_closure
    TC.add_edges_from(edges)
  File "/home/vdshk/.local/lib/python3.8/site-packages/networkx/classes/multigraph.py", line 550, in add_edges_from
    for e in ebunch_to_add:
  File "/home/vdshk/.local/lib/python3.8/site-packages/networkx/algorithms/dag.py", line 575, in <genexpr>
    edges = ((v, w) for u, w in nx.edge_dfs(G, v))
ValueError: too many values to unpack (expected 2)
```

### Expected Behavior
```python
>>> import networkx as nx
>>> g = nx.MultiDiGraph([(0, 1), (1, 2)])
>>> print(nx.transitive_closure(g).edges())
[(0, 1), (0, 2), (1, 2)]
```

### Environment
Python version: `3.8.10`
NetworkX version: `2.6.2`

### Proposal

I think this function needs to be refactored.
Maybe it should be done in #5042 PR.
What do you think?

**Comment by rossbar** [comment]
+1 for adding support for multigraphs. I'd recommend putting it in a separate PR though rather than adding it to #5042 - IMO it would make it easier to review if there were separate PRs for each proposal. If you want to work on it before #5042 is in, you can always open a new PR that depends on #5042 (i.e. originates from a commit on that branch) and just mark it as draft for now.
